### PR TITLE
Command wrapper function

### PR DESCRIPTION
**Pull request by svenkeidel**
This pull-request adds a customizable `flycheck-command-wrapper-function` and `flycheck-executable-find` function. The purpose of those is discussed in [this](https://github.com/flycheck/flycheck/issues/629) feature request. I'm open for suggestions and improvements.

Best,
Sven

Edit (lunaryorn): Connects to #629

### CODE REVIEW

**Comment by svenkeidel** [comment]
BTW, I plan to publish the NixOS specific code that gets hooked into the `flycheck-command-wrapper-function` in a separate repository.

File: flycheck.el [hunk 1]
```diff
@@ -4489,6 +4529,9 @@ symbols in the command."
     (condition-case err
         (let* ((program (flycheck-checker-executable checker))
                (args (flycheck-checker-substituted-arguments checker))
+               (command (funcall flycheck-command-wrapper-function
```
Comment (lunaryorn) [inline flycheck.el]: I've thought about this again.  Why don't we pass `(cons program args)` to the wrapper function, whose default value would then be `#'identity`?  That'd make for a much nicer API in my opinion, what do you think?

**Comment by lunaryorn** [comment]
I think we can have a better API here, see my comment on the diff.

Besides, I'm very sorry for the delay, I'm pretty busy these days with work, life, and trying Spacemacs :blush:

**Comment by boothead** [comment]
Hi All,

I've just found this thread while refreshing my dev environment for spacemacs on nix. Is there anything I can do to help here?

The current version of flycheck in spacemacs is still pretty confused with nix-shell. It doesn't find dependencies of the current project even when emacs is launched from within nix-shell.

Please feel free to ping me if there's anything that you'd like me to test!

**Comment by lunaryorn** [comment]
@boothead I don't think that this change still needs testing.  What it does need, though, is [an improved API](https://github.com/flycheck/flycheck/pull/752#discussion_r42037713), and probably also some unit tests… and ultimately some time from me to review and merge :relaxed:

**Comment by svenkeidel** [comment]
@boothead, I have implemented the search for executables in nix sandboxes [here](https://github.com/svenkeidel/emacs.d/blob/master/nixos.el).

@lunaryorn Yes, I think you are right. I will change the code.

File: flycheck.el [hunk 2]
```diff
@@ -417,6 +417,45 @@ If set to nil, do not show error tooltips."
   :package-version '(flycheck . "0.25")
   :risky t)
 
+(defcustom flycheck-command-wrapper-function
+  #'identity
+  "Wraps checker commands and its arguments before it gets executed.
+
+The function is usefull to wrap a checker command, if the checker
+is only available in sandbox-like enviroments.  The default value
+does not modify the command or arguments.  In NixOS, checkers are
```
Comment (lunaryorn) [inline flycheck.el]: I understand that it's important to you, but could you nontheless please shorten the references to NixOS in this docstring to just a simple example?  See, from my point of view NixOS is still just a niche product, and there are many other things this wrapper function can be used for (I'm thinking of `bundle exec` or Cabal sandboxes, etc.).

I'd like to keep the docstring short and to the point.  Instructions or references to very specific systems shouldn't get much place here.

File: flycheck.el [hunk 3]
```diff
@@ -4694,9 +4738,9 @@ variable symbol for a syntax checker."
           (executable (if current-prefix-arg
                           nil
                         (read-file-name "Executable: " nil default-executable
-                                        nil nil #'executable-find))))
+                                        nil nil #'flycheck-executable-find))))
```
Comment (lunaryorn) [inline flycheck.el]: Ah, I just noticed that you need to remove `#'` here.  `flycheck-executable-find` is not a function on its own, but just a variable referring to a function.  Hence the function quote is wrong at this place.

Comment (svenkeidel) [inline flycheck.el]: I'm sorry about that one. I'm spoiled by the type checker of Haskell :smile:

**Comment by boothead** [comment]
@svenkeidel That package looks very handy, thanks. I'm going to try pulling this branch into my spacemacs layer to see if I can get it up and running.

**Comment by lunaryorn** [comment]
@svenkeidel Merged with some smaller cleanups, plus tests and docs.  Thank you for this contribution!

/cc @boothead @purcell

**Comment by lunaryorn** [comment]
@svenkeidel Just being curious, how's your Flycheck package for NixOS going?  Did you already publish it somewhere?

**Comment by svenkeidel** [comment]
I'm currently working on getting it merged [here](https://github.com/travisbhartwell/nix-emacs).

**Comment by lunaryorn** [comment]
@svenkeidel Great!  Please ping me when you're done, and I'll add the package to http://www.flycheck.org/extensions.html

**Pull request closed without merging**